runtime/src/accounts_db.rs

```diff
@@ -773,15 +780,29 @@ impl AccountsDB {
 
         self.report_store_stats();
 
+        let mut key_create = Measure::start("key_create");
+        let mut pubkeys: Vec<Pubkey> = Vec::new();
+        {
+            let mut uncleaned_keys = self.uncleaned_keys.lock().unwrap();
+            let max_slot = max_clean_root.unwrap_or(u64::MAX);
+            let slots: Vec<Slot> = uncleaned_keys.keys().cloned().collect();
+            for slot in slots {
+                if slot < max_slot {
```

Comment:
In the old code, we only care about the pubkeys for rooted slots, so we could check filter these slots by the `AccountsIndex.roots_tracker.uncleaned_roots` set as well. One thing to note is that @ryoqun  noticed quite a lot of contention on that `uncleaned_roots` lock already: https://github.com/solana-labs/solana/pull/14578#issuecomment-760061156, so it would be good to extract it into a `DashSet`